kmip/services/server/engine.py

```diff
@@ -1501,6 +1504,55 @@ def _process_activate(self, payload):
         return response_payload
 
     @_kmip_version_supported('1.0')
+    def _process_revoke(self, payload):
+        self._logger.info("Processing operation: Revoke")
+
+        revocation_code = None
+        if payload.revocation_reason and \
+           payload.revocation_reason.revocation_code:
+            revocation_code = payload.revocation_reason.revocation_code
+        else:
+            raise exceptions.InvalidField(
+                "revocation reason code must be specified"
+            )
+
+        if payload.unique_identifier:
+            unique_identifier = payload.unique_identifier.value
+        else:
+            unique_identifier = self._id_placeholder
+
+        managed_object = self._get_object_with_access_controls(
+            unique_identifier,
+            enums.Operation.REVOKE
+        )
+        object_type = managed_object._object_type
+        if not hasattr(managed_object, 'state'):
+            raise exceptions.IllegalOperation(
+                "An {0} object has no state and cannot be revoked.".format(
+                    ''.join(
+                        [x.capitalize() for x in object_type.name.split('_')]
+                    )
+                )
+            )
+
+        # TODO: (1) need to set Compromise Date attribute or Deactivation Date
+        # attribute, (2) The revoke (KMIP1.2 4.20)  doesn't say anything about
+        # the state transition permission, but the state (KMIP1.2 3.22) diagram
+        # does show that not all states can transit to "Deactivated" or
+        # "Compromised". So what's the right way?
```

Comment:
Yeah, the info for this is spread around Sections 3.22 and 4.20. This is how the state transition should go.

If the reason code is `KEY_COMPROMISE`:
`PRE_ACTIVE` -> `COMPROMISED`
`ACTIVE` -> `COMPROMISED`
`DEACTIVATED` -> `COMPROMISED`
`DESTROYED` -> `DESTROYED_COMPROMISED`

If the reason code is *not* `KEY_COMPROMISE`:
`ACTIVE` -> `DEACTIVATED`

If someone tries to revoke an object that is not in the `ACTIVE` state, for a reason that is *not* a compromise, that should be an error (`IllegalOperation` seems fine).